### ISSUE TITLE
PoisonPill stops actors without waiting for the mailbox to be processed

### ISSUE THREAD

**Issue by SkinyMonkey** [asynkron/protoactor-go#391]
"Poison will tell actor to stop after processing current user messages in mailbox"
This is what the documentation says.

When in fact this is what happens:

```
func (ctx *actorContext) defaultReceive() {
	if _, ok := ctx.Message().(*PoisonPill); ok {
		ctx.Stop(ctx.self)
		return
```

The actor is stopped without waiting for the mailbox to be processed.
Unless i missed something?

I did a workaround for now .. but i'd be happy to remove it.

**Comment by rogeralsing** [comment]
PoisonPill is a user message and will allow all messages _before_ the pill on the mailbox to be fully processed, once the pill is consumed, the actor is stopped.

Meaning anything after the pill will be discarded.

**Comment by SkinyMonkey** [comment]
Ok, that make sense, maybe the comment could be changed to avoid any confusion?
"current user messages in mailbox" => that made me think that somehow the mailbox would stop taking messages and consume every messages left in the mailbox before the actor dies.
Which could make sense in some cases i think:
I have a pool of actors talking to each other and when this pool is cancelled i just want them to stop talking so I can stop them without deadletters.

I do have an implementation based on a middleware and an atomic but i'm wondering if someone had a better idea.

**Comment by alexeyzimarev** [comment]
The point is that if your actor system keeps receiving messages from the outside, you will end up with dead letters anyway. Poison pills only allow you to say "ok, I am letting the actor to complete all unfinished business and then it must stop" but it won't tell anyone else to stop sending messages to that actor.

Presumably, you need to stop accepting messages from the outside and _then_ distribute all the poison pills, like in a good old doomsday cult :)

**Comment by SkinyMonkey** [comment]
It make sense @alexeyzimarev, thank you. But i've kept the middleware + atomic way:

Based on your idea I designed a set of theoretical plugins/mixins to solve this.
Every parent part of the abort has to count its children.
Then on abort i tell every children to stop listening, with a special message.
They then have to tell their parent they received the message, the parent count the children and when every children has told them "i stopped listening" they tell their parent etc.
The root entity can then stop everything cleanly.
But! At lot of messaging is going on, plus all the messages before the "stop listening" message have to be treated for nothing, as its a user message...

The atomic + middleware design, that i called a "faucet", have the advantage of stopping ANY discussion when it's closed, without doing any further message treatment and the implementation is way simpler/smaller.
Basically I propagate a middleware to every children of a root entity.
This middleware checks atomically if an int32 is in a state or the other (open/closed) on send and receive of a message.
On abort, the root entity close the faucet through a method and sleeps a bit to be sure that the atomic was closed (we don't care we're in an abort case) and then stops.

I know it brakes both golang's mantra "communicate to share state" and the actor model, but ... it works, it's quick and it's a simple design.
If someone think i'm unknowingly shooting myself in the foot with that second design, please tell me!

**Comment by mec07** [comment]
@SkinyMonkey I'm having the same problem. Thanks for raising the issue and describing your solution. Can you share your middleware code? Thanks!

**Comment by rogeralsing** [comment]
PoisonPill only consumes the mailbox for the actor receiving the pill, for all child and descendants of that actor, they will be stopped using Stop

https://github.com/asynkron/protoactor-go/blob/dev/actor/actor_context.go#L608-L617

So the stop is graceful in the sense that all actors in the tree are properly stopped once the poison-pill is consumed.
It is not graceful in the sense that all child actor mailbox messages are consumed.

It would be pretty easy to make it so. e.g. if poisoned, all children are also poisoned.
But that would be a pretty significant change in behavior.

Thoughts around this?

**Comment by mec07** [comment]
Personally, I would go for poisoning all children recursively down the tree. `PoisonFuture(pid).Result()` would then involve waiting for the future response to propagate back up the tree too. Otherwise it's not really a graceful shutdown. Graceful shutdown is really important and if it's not there then the users will have to reinvent the wheel over and over again. I was literally just thinking about creating my own custom poison pill which does this for my implementation.

**Comment by mec07** [comment]
I wanted to do something like this in the middleware, but I can't access the `context.Children()` or do `context.RequestFuture` from a receive middleware, because it gets the `ReceiveContext` type instead of a normal `Context`. So I've had to resort to a rather crude implementation using the actor `Receive` function.

Just in case it's helpful to anyone else, here is a very crude implementation of a graceful shutdown using a custom message (I really hope that this can be replaced by a neater solution from the `protoactor-go` package as it's quite messy):
```
package supervisor

const maxShutdownDuration = time.Minute

type Shutdown struct{}

func (a *MyActor) Receive(context actor.Context) {
        switch msg := context.Message().(type) {
        case Shutdown:
                var futures []*actor.Future
                for _, child := range context.Children() {
                         future := context.RequestFuture(child, msg, maxShutdownDuration)
                         futures = append(futures, future)
                }
                for _, future := range futures {
                        err := future.Wait()
                        if err != nil {
                                log(err)
                        }
                }
                context.Respond(msg)
                context.Stop(context.Self())
        }
}
```
You've got to copy and paste the `Shutdown` case into the `Receive` function for each actor you have in your system  (well, create a function with the code block so it's just one line).

So, starting and stopping your actor system like this looks like:
```
// Start up actor system
props := actor.PropsFromProducer(func() actor.Actor {
        return &supervisor.MyActor{}
})
pid := system.Root.Spawn(props)

// Graceful shutdown of actor system (would be sensible to put this into a defer)
if err := system.Root.RequestFuture(pid, supervisor.Shutdown{}, time.Minute).Wait(); err != nil {
        log(err)
}
```
